### DISCUSSION TITLE
Sync DAVx5 CALDAV Client Cache to Xadikos

### DISCUSSION THREAD

**Discussion by lamyergeier** [jelmer/xandikos#178] [Q&A]
I had installed CALDAV server (Radicale) on my old laptop, but now I don't have access to that machine. I had synced calendars hosted by the caldav server to my android using DAVx5. 

Now, on my new machine I have installed a CALDAV server (Xandikos), and I would like to sync the existing CALDAV calendars on Android that were synced to Radicale be synced to Xandikos. 

Essentially the base url of each calendar needs to be updated, from http://192.168.178.22:5232/nikhil to http://192.168.178.33:8899/users/calendars. 

Any suggestions?

![Screenshot_20220708-140148_DAVx](https://user-images.githubusercontent.com/42092626/177989798-1864fd23-d9ee-4ed4-bb46-f015940c8913.jpg)

---

# Steps that I tried

1. I created folder `mkdir  -p /Users/lamyergeier/dav/nikhil/calendars`
2. I launched Xadikos with 192.168.178.22 and port 5232: `/Users/lamyergeier/.local/bin/xandikos -d /Users/lamyergeier/dav -l 192.168.178.22 -p 5232`

## Observation

On the browser http://192.168.178.22:5232 doesn't load the web interface! and it shows Firefox can’t establish a connection to the server at 192.168.178.22:5232.

**Comment by jelmer** [comment]
The question about DAV5X is probably best asked in a DAV5X forum. Let's focus on the fact that you can't actually reach Xandikos here.

What output do you get from Xandikos itself? Does it say it's listening on http://192.168.178.22:5232/ ? Can you reach http://192.168.178.22:5232/ using e.g. curl on the machine where you're running Xandikos.

**Comment by lamyergeier** [comment]
```
$ /Users/lamyergeier/.local/bin/xandikos -d /Users/lamyergeier/dav -l 192.168.178.22 -p 5232
INFO:root:Listening on 192.168.178.22:5232
WARNING:root:aiohttp-openmetrics not found; /metrics will not be available.
ERROR:asyncio:unhandled exception during asyncio.run() shutdown
task: <Task finished name='Task-1' coro=<_run_app() done, defined at /Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web.py:287> exception=OSError(49, "error while attempting to bind on address ('192.168.178.22', 5232): can't assign requested address")>
Traceback (most recent call last):
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web.py", line 514, in run_app
    loop.run_until_complete(main_task)
  File "/usr/local/Cellar/python@3.10/3.10.4/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web.py", line 413, in _run_app
    await site.start()
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web_runner.py", line 121, in start
    self._server = await loop.create_server(
  File "/usr/local/Cellar/python@3.10/3.10.4/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1505, in create_server
    raise OSError(err.errno, 'error while attempting '
OSError: [Errno 49] error while attempting to bind on address ('192.168.178.22', 5232): can't assign requested address
Traceback (most recent call last):
  File "/Users/lamyergeier/.local/bin/xandikos", line 31, in <module>
    sys.exit(main(sys.argv))
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/xandikos/__main__.py", line 28, in main
    return main(argv)
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/xandikos/web.py", line 1359, in main
    web.run_app(app, port=listen_port, host=listen_address, path=socket_path)
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web.py", line 514, in run_app
    loop.run_until_complete(main_task)
  File "/usr/local/Cellar/python@3.10/3.10.4/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web.py", line 413, in _run_app
    await site.start()
  File "/Users/lamyergeier/.local/pipx/venvs/xandikos/lib/python3.10/site-packages/aiohttp/web_runner.py", line 121, in start
    self._server = await loop.create_server(
  File "/usr/local/Cellar/python@3.10/3.10.4/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1505, in create_server
    raise OSError(err.errno, 'error while attempting '
OSError: [Errno 49] error while attempting to bind on address ('192.168.178.22', 5232): can't assign requested address
```

@jelmer I was wondering that since the CALDAV server uses 2 way sync, I could simply sync from the client (DAVx5) by making the base URL of the calendars in the client same as that on the Xandikos server.

**Reply by jelmer** [reply]
It looks like there's probably already something listening on that port.  Radicale perhaps?

> @jelmer I was wondering that since the CALDAV server uses 2 way sync, I could simply sync from the client (DAVx5) by making the base URL of the calendars in the client same as that on the Xandikos server.

That could work, but calendars in Xandikos are git repositories so you'll need to create those using git rather than using mkdir.

**Comment by lamyergeier** [comment]
@jelmer I started xandikos using ip `/Users/lamyergeier/.local/bin/xandikos -d /Users/lamyergeier/dav -l 192.168.178.33 -p 5232`

Can you please tell me how can I check that which process is listening on IP 192.168.178.33? By this I can check the process listening on  192.168.178.22. 

I don't have Radicale installed anymore as I don't have that machine. As per my estimation this IP address should be free.


I tried
  -  lsof -i 4tcp
  - netstat

but they don't show the IP on which Xandikos is listening!

**Reply by jelmer** [reply]
"netstat -lp" might help you on unix machines - but this is really an OS-specific question than related to Xandikos.

**Reply by lamyergeier** [reply]
@jelmer Thank you! I have now connected xandikos to the same IP and port to which the radicale was connected on the old device.

My directory structure is as follows

```
$ pwd
/Users/lamyergeier/dav
$ tree
.
├── nikhil
│   └── calendars
└── user
    ├── calendars
    ├── contacts
    └── inbox

6 directories, 0 files
```

- I am trying to sync from DAVx5 but nithing sync's and I get the following error:

```
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:17:49:10 +0000] "PROPFIND /nikhil/6229e809-b885-b85b-d83a-a799bfa7040e/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:17:49:11 +0000] "REPORT /nikhil/6229e809-b885-b85b-d83a-a799bfa7040e/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
```

Here, 
- `6229e809-b885-b85b-d83a-a799bfa7040e` is the calendar folder name.
- 192.168.178.21 is the IP address of Android

# About the Web interface

Even though I have created folder users and nikhil under the base folder, the web interface just shows `user`, the default principal!

![image](https://user-images.githubusercontent.com/42092626/178045289-f2847b01-fc2c-42e2-bab0-96d6360aef6e.png)

Why is it not showing nikhil as well? Also, why the two way sync not working?

Also, as far as I understand git is not mandatory for the functioning of CALDAV server right? That's only needed when we want to log changes to history. So, it should not be an issue.

**Reply by jelmer** [reply]
You need to tell Xandikos what you want the principal to be; pass in --current-user-principal=/nikhil/

**Reply by lamyergeier** [reply]
Thanks I did this. The calendars are not yet synced, but I get the following log 2 lines for each calendar

```
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:05 +0000] "PROPFIND /nikhil/6229e809-b885-b85b-d83a-a799bfa7040e/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:05 +0000] "REPORT /nikhil/6229e809-b885-b85b-d83a-a799bfa7040e/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:06 +0000] "PROPFIND /nikhil/f69ef521-c33a-43bd-eb2c-86e14d0fa447/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:06 +0000] "REPORT /nikhil/f69ef521-c33a-43bd-eb2c-86e14d0fa447/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:06 +0000] "PROPFIND /nikhil/d994d01f-7150-c6d0-6d3d-b6c4d9ff8b30/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:06 +0000] "REPORT /nikhil/d994d01f-7150-c6d0-6d3d-b6c4d9ff8b30/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:07 +0000] "PROPFIND /nikhil/e5b93e12-3aa4-b252-4bfe-ad683ec3bbd3/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:07 +0000] "REPORT /nikhil/e5b93e12-3aa4-b252-4bfe-ad683ec3bbd3/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:07 +0000] "PROPFIND /nikhil/e8747b12-ac9d-d3f1-9ba6-2c71b8c2dd08/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:07 +0000] "REPORT /nikhil/e8747b12-ac9d-d3f1-9ba6-2c71b8c2dd08/ HTTP/1.1" 404 219 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:07 +0000] "PROPFIND /nikhil/c0c52cd5-82dd-f7b0-e193-4f63458713e9/ HTTP/1.1" 207 387 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
INFO:aiohttp.access:192.168.178.21 [08/Jul/2022:19:55:08 +0000] "PUT /nikhil/c0c52cd5-82dd-f7b0-e193-4f63458713e9/fa47e547-57f7-48fa-9a17-53c33954c623.ics HTTP/1.1" 404 260 "-" "DAVx5/4.2.1-ose (2022/05/08; dav4jvm; okhttp/4.9.3) Android/11"
```

**Reply by jelmer** [reply]
Have you actually created the calendar on the xandikos side with git?
